cmd/kube-apiserver/app/server.go

```diff
@@ -181,8 +182,12 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.CertDirectory, "cert-dir", s.CertDirectory, "The directory where the TLS certs are located (by default /var/run/kubernetes). "+
 		"If --tls-cert-file and --tls-private-key-file are provided, this flag will be ignored.")
 	fs.StringVar(&s.APIPrefix, "api-prefix", s.APIPrefix, "The prefix for API requests on the server. Default '/api'.")
-	fs.MarkDeprecated("api-prefix", "--api-prefix is deprecated and will be removed when the v1 API is retired")
-	fs.StringVar(&s.StorageVersion, "storage-version", s.StorageVersion, "The version to store resources with. Defaults to server preferred")
+	fs.MarkDeprecated("api-prefix", "--api-prefix is deprecated and will be removed when the v1 API is retired.")
+	fs.StringVar(&s.StorageVersion, "storage-version", s.StorageVersion, "The version to store the legacy v1 resources with. Defaults to server preferred")
+	fs.MarkDeprecated("storage-version", "--storage-version is deprecated and will be removed when the v1 API is retired. See --storage-versions instead.")
+	fs.StringVar(&s.StorageVersions, "storage-versions", s.StorageVersions, "The versions to store resources with."+
+		"Different groups may be stored in different versions. Specified in the format \"group1/version1,group2/version2...\"."+
+		"Defaults to server preferred.")
```

Comment:
Why do changes require the upgrade script?  You can still run even if you set this, you might just not realize that you still have objects in the older version until after the old version is no longer supported, in which case you're broken.  Which is bad, for sure.

The one advantage to leaving it as the default is in a rolling upgrade you get the correct behavior.